feat(vertical-navigation): Add the Vertical Navigation pattern

mturley opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:

Creating the new Vertical Navigation component.

<!-- Why are these changes necessary? -->
**Why**:

To bring patternfly-react one step closer to feature parity with patternfly-ng

<!-- How were these changes implemented? -->
**How**:

New components, factoring things to be as reusable as possible

<!-- feel free to add additional comments -->

Storybook: https://mturley.github.io/patternfly-react/?selectedKind=Vertical%20Navigation&selectedStory=Items%20as%20JSX&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs

mturley commented:
@ohadlevy, Serena told me that I should ping you when I had this pushed up.  It's not quite done yet, i'm still in the process of refactoring things.  I pushed my partial code so I can head home, but I plan to wrap up most of this tonight.

priley86 commented:
@mturley It looks like it's still a bit early, but I think this is the right idea. I.m.o. the only component state we’ll need is the state which indicates whether the secondary/tertiary nav drawers are open, everything else can be indicated by props from the application/redux store (including what nav items are active). I like the approach w/ rendering the nav items as lists.

Look forward to reviewing it further once you are ready here...

ohadlevy commented:
/cc @amirfefer

mturley commented:
Sorry for the delay on this, I've had a few distractions.  I just pushed a few more commits that complete the basic structure of this and include some refactoring I did to make the primary/secondary/tertiary levels all driven by one VerticalNavigationItem (the markup was similar enough at each level that the differences are logically laid out in the render function without foo much fuss).  I think it's pretty clean.

Now I just need to finish getting any react-specific CSS tuned up, and address my last few TODO comments.  Should be good to go on Monday.

My next PR will be at a faster pace-- I've had some new-hire tasks and other one-time distractions these first few weeks.

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigation.js:
```diff
@@ -0,0 +1,252 @@
+      activeSecondary,
+      children,
+      topBannerContents,
+      notificationDrawer /* TODO notification drawer components? */,
```
I would leave this out and allow the application to add it separately. The drawer trigger should be added to the topBannerContents and notification drawer would be part of the app content.

mturley replied:
Well, that is sort of what I'm doing here already. The notification drawer is a child of the top banner, so it has to be rendered by this component (React doesn't allow external DOM modifications). So we just accept a prop `notificationDrawer` that is just the entire node tree to be rendered in that container.  The application can pass a whole other component as that prop.

jeff-phillips-18 replied:
I don't believe notificationDrawer needs to be a child of the top banner.

jeff-phillips-18 replied:
If it is part of the vertical navigation, it does buy you the correct top positioning and height otherwise it is not necessary, the application would then need to specify those values. We could handle that in the notification drawer component.

Just seems better to keep them separated to me. I'm OK either way though.

mturley replied:
I see. Okay. It was a child of the top banner in the pf-ng code and the PF core example markup. If we don't do that, I feel like we should make that change upstream maybe?

jeff-phillips-18 replied:
there is no upstream change necessary though. You can have it either way. There is just the preset position and height setting gained by having the drawer-pf a child of navbar-pf-vertical

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigation.js:
```diff
@@ -0,0 +1,252 @@
+    const { hideTopBanner } = this.state;
+    this.setState({
+      hideTopBanner: !hideTopBanner,
+    });
```
This will need to notify the application as well. The main body content needs to respond to the collapse as well.

mturley replied:
Good point. I'll add a handler that can be passed in for that.

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigationItem.js:
```diff
@@ -0,0 +1,215 @@
+          // This class is confusingly named, but the logic is more readable.
+          'mobile-secondary-item-pf':
+            mobileItem && depth === 'primary' && showMobileTertiary,
+          // I don't know, that's just how this stuff was in patternfly-ng...
```
Yeah, could have been named better but basically it is used to handle mobile navigation where only one of the menus (primary, secondary, or tertiary) are shown at a time (the navigation is drill down in mobile mode).

mturley replied:
Makes sense. I'll change this comment to be a little less confused, haha.

mturley commented on src/components/VerticalNavigation/VerticalNavigationItem.js:
```diff
@@ -0,0 +1,215 @@
+                />
+                <span>{title}</span>
+              </div>
+              <ListGroup>
```
@priley86 , here is the cause of the <li>-inside-<li> warnings we are getting. This ListGroup sits inside a ListGroupItem, and renders ListGroupItem children.

jeff-phillips-18 replied:
I think you need to add `componentClass="ul"` here.

priley86 commented:
@mturley do you mind rebasing w/ current master? I think something in index.js is misaligned. Getting the following:
```
ERROR in ./src/index.js
Module not found: Error: Can't resolve './components/VerticalNavigation' in 
```

jeff-phillips-18 commented:
@mturley Looks like we need index.js in components/VerticalNavigation

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigation.js:
```diff
@@ -0,0 +1,252 @@
+              'show-mobile-nav': showMobileNav,
+            })}
+          >
+            <ListGroup>
```
I think you need to add `componentClass="ul"` here.

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigationItem.js:
```diff
@@ -0,0 +1,215 @@
+
+VerticalNavigationItem.defaultProps = {
+  trackActiveState: true,
+  trackHoverState: true,
```
why are the defaults true here? Aren't these indicating the item is active or hovered on? Only one item at most should ever be active, same as hover.

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigationItem.js:
```diff
@@ -0,0 +1,215 @@
+          [`${nextDepth}-nav-item-pf`]:
+            depth !== 'tertiary' && children && children.length > 0,
+          active: trackActiveState, // This is the only class we have at the tertiary depth.
+          'visible-xs-block': trackActiveState, // TODO is trackActiveState the right thing here, and is it named right?
```
Not sure why 'visible-xs-block' set on active items (or ever).

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigationItem.js:
```diff
@@ -0,0 +1,215 @@
+      handleTertiaryBlur,
+      handlePrimaryClick,
+      handleSecondaryClick,
+      handleTertiaryClick,
```
Why would an item have 3 levels of click handlers? Any one item would only ever be at one level right? Shouldn't it just have a handleClick, handleHover, handleBlur?

Not really sure why any item needs the handleHover or handleBlur callbacks. Shouldn't the hover state be maintained here? Also, you will find the need to have delays for the hover/blur handling when showing/hiding submenus.

mturley commented:
I'm almost done with this, I squashed all the commits so far and I'll squash again after I push the rest.

jeff-phillips-18 commented:
@mturley Would it make sense to create this component w/o secondary or tertiary navigation support to begin with? This gets us partially there (and enough of the way for those w/o secondary menus) and able to try this out before adding the hover/pinned features.

Thoughts?

priley86 commented:
@mturley @jeff-phillips-18 I guess my initial thought after reviewing it a bit more is why have `<VerticalNavigationItem />` 's listed and then _also_ have the `mockNavItems` array. It just seems strange to do lookups to me (and I would you think you would have one or the other to avoid prop merges, etc). But I'm fine with iterating on this a bit more and seeing what others think too...
 
Good job putting this together so quickly @mturley. It's not a simple component.

priley86 commented on src/components/VerticalNavigation/VerticalNavigation.js:
```diff
@@ -0,0 +1,313 @@
+      collapsedSecondaryNav: false,
+      collapsedTertiaryNav: false,
+    };
+    this.handleNavBarToggleClick = this.handleNavBarToggleClick.bind(this);
```
We can use the bind helpers in helpers.js here now:
https://github.com/patternfly/patternfly-react/blob/master/src/common/helpers.js

priley86 replied:
(see Wizard examples)

priley86 commented on src/components/VerticalNavigation/VerticalNavigation.stories.js:
```diff
@@ -0,0 +1,53 @@
+  () => {
+    return (
+      <VerticalNavigation secondaryCollapsed={false} tertiaryCollapsed>
+        <VerticalNavigationItem title="Item 1" />
```
This is a nit but can we expose the API as `VerticalNavigation.Item` ... see Wizard examples...it's just a consistent thing with the other complex components (both in PF React and React Bootstrap) where we have a lost of sub components.

priley86 commented:
@mturley apologies ignore my last comment about `mockNavitems` - I misread the code ;)

priley86 commented:
@mturley once you get closer on this, you may want to test w/ React Router. I don't know that we should necessarily assume that VerticalNavigationItem will be a direct child of VerticalNavigation (depending on how we set up the React Router Dom). Reason being is - typically you will see `<Route>` as a wrapper to any routable component which accepts a match property (as a function). The match property typically determines active state (based on the active route).

I started some examples of this here prior, but this is missing all of the other vert. nav features you've started in the implementation here. I would say just test this sooner than later ;->

https://github.com/patternfly/patternfly-react-demo-app/blob/flowjs/src/components/Nav/VerticalNav.js

https://github.com/patternfly/patternfly-react-demo-app/blob/flowjs/src/components/Nav/RouteNavItem.js

priley86 commented on src/components/VerticalNavigation/VerticalNavigationItem.js:
```diff
@@ -0,0 +1,236 @@
+
+    const handleHover = {
+      primary: handlePrimaryHover,
+      secondary: handleSecondaryHover,
```
Re-reading this bit... I almost wonder if splitting out Secondary and Tertiary into their own components makes more sense (as it seems it would split the components a bit better as far as props/handlers/etc, i.e.:

```
<VerticalNavigation secondaryCollapsed={false} tertiaryCollapsed>
   <VerticalNavigation.Item title="Item 1">
      <VerticalNavigation.SecondaryItem title="Item 1-A"> 
         <VerticalNavigation.TertiaryItem title="Item 1-A-i">
      </VerticalNavigation.SecondaryItem> 
   <VerticalNavigation.Item>
</VerticalNavigation>         
```
It seems like this may split the logic a bit better and make it a bit easier to compose...
@jeff-phillips-18 thoughts?

priley86 replied:
I may be over simplifying a bit though... not sure.

mturley commented:
@jeff-phillips-18 , sorry I didn't notice your comment earlier about doing primary-only. I think I will have everything working shortly, but if there are any further issues with secondary/tertiary stuff I will disable them for now and get this merged.

mturley commented:
@jgiardino how come you removed this from In Progress on the board?

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigation.js:
```diff
@@ -0,0 +1,379 @@
+    const item = deepestOf(primary, secondary, tertiary);
+    if (inMobileState) {
+      if (item.subItems && item.subItems.length > 0) {
+        this.updateMobileMenu(primary, secondary, tertiary); // TODO figure out what this did in ng
```
In mobile state, for primary items with secondary children, a click results in showing the secondary nav menu. Same for secondary items with tertiary children showing the tertiary nav menu.

mturley replied:
Yep, figured that out earlier and I'm still getting around to cleaning up these old comments.

jeff-phillips-18 commented:
We determined in testing vertical navigation usability that the hover and blur responses should be delayed. The problem is that when you hover an item and its submenu is shown, when moving the pointer to the submenu, if you move off of the hover item the submenu is removed and you have to start over again (very annoying).

In the Patternfly Jquery and in the Angular Patternfly implementations, there is a hover delay of 500ms before showing the submenu and a blur delay of 700ms before hiding the submenu.

mturley commented:
@jeff-phillips-18, that is implemented already. See https://github.com/patternfly/patternfly-react/pull/88/files#diff-4ff0226849a5e26980c99b15c1d255d4R466, https://github.com/patternfly/patternfly-react/pull/88/files#diff-ada5f5d653de82f6b79dbc7b02da8f6dR116, https://github.com/patternfly/patternfly-react/pull/88/files#diff-ada5f5d653de82f6b79dbc7b02da8f6dR148

mturley commented:
This is 99% done, i am just having trouble getting the storybook build to work. When I finish troubleshooting it I will push for a final round of review.

jeff-phillips-18 commented:
When running the storybook link above, I'm not seeing the blur delay.

mturley commented:
What storybook link? I have not gotten a successful rawgit storybook build of my version of vertical nav working yet.

mturley commented:
If you build my branch locally it works.

mturley commented:
My last refactor is still allllmost done, I'll push it tonight if I can. Aiming to have a big review of all this stuff tomorrow or Wednesday if possible, @jeff-phillips-18 and @priley86

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigation.stories.js:
```diff
@@ -0,0 +1,142 @@
+  `VerticalNavigation usage example with items passed as objects, driven by a mock file.`,
+  () => {
+    return (
+      <MockPositionFixed>
```
Add MockPositionFixed to the propTablesExclude property.

jeff-phillips-18 replied:
Add subcomponents to propTables property.

mturley commented:
Serena reminded me that I had not yet put a storybook link in here, so I'll add it to the description above.

mturley commented:
Only router testing and unit testing remain here. Thank you guys for your patience while I perfect this thing!

priley86 commented on src/common/helpers.js:
```diff
@@ -1,7 +1,79 @@
+  contained here. When present, props are used instead. If you provide these special props,
+  be sure to also provide corresponding callbacks/handlers to keep them updated.
+*/
+export const controlled = (stateTypes, defaults = {}) => WrappedComponent => {
```
+1... again, great job on adding this HOC!

mturley replied:
I'll put this in its own file. I think eventually though, I want to get it merged into recompose :P but we'll see if they still disagree when I refactor it to use `withStateHandlers`. Future enhancement, though, not worrying about that now.

priley86 commented on src/common/patternfly.js:
```diff
@@ -1,7 +1,9 @@
+import Break from 'breakjs';
 
 const patternfly = window.patternfly;
+const layout = Break({ mobile: 0, ...patternfly.pfBreakpoints });
```
surprised we don't have `0` in Core yet? Maybe that would be a good addition in the future... not sure if anyone is opposed to this?

priley86 commented on src/components/VerticalNavigation/constants.js:
```diff
@@ -0,0 +1,105 @@
+// WARNING: HACK! HAAAACK HACK HACK HACK WARNING THIS IS A HACK.
+// We only use this to apply magic body classes when the component is used in uncontrolled mode. TODO it's a prop now
+// And only for consistency-- the better solution is to manage these classes yourself in the application.
+const getBodyContentElement = () => {
```
yep - remember discussing this one w/ you a bit back. Thanks for calling it out in the comments - ideally PF Next does not do these sort of things on the `<body>` or `<html>` tags. React's context typically renders inside the top most element after `<body>` (typically the `<div>` beneath `<body>`) and the state tree/redux store/application begins there...  so for that reason, in the future I would propose we may consider a `Layout` component possibly (and use that to handle this sort of layout css state)? I think the way Material UI does this is very nice, in the way that it renders Themes and ThemeProviders at the top most level:
https://material-ui-next.com/customization/themes/

Any classes that render on the `<body>` tag are highly suspect to me. That being said...having `VertNav` as the top most component beneath Redux/Application seems to be OK for now and matches our current PF3 specs...

jeff-phillips-18 replied:
Just to be clear, this is not the `<body>` element. This is the main content area for the application.

priley86 replied:
ahh - apologies, was conflating a different thing i think ;). thanks!

priley86 commented on src/components/VerticalNavigation/index.js:
```diff
@@ -0,0 +1,31 @@
+VerticalNavigation.Item = Item;
+VerticalNavigation.Brand = Brand;
+VerticalNavigation.Collapse = Collapse;
+VerticalNavigation.IconBar = IconBar;
```
very nice on the breakdown here!

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigationBrand.js:
```diff
@@ -0,0 +1,39 @@
+  );
+};
+
+VerticalNavigationBrand.propTypes = {
```
Would be useful to add storybook descriptions.

mturley replied:
Definitely on the short list of remaining things.  And as for that little bit of CSS, maybe it should be in pf core but I wasn't sure where that style was coming from in patternfly-ng, it appears to render that way but I didn't see that CSS anywhere. Still needs a little investigating.

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigationMasthead.js:
```diff
@@ -0,0 +1,68 @@
+  );
+};
+
+BaseVerticalNavigationMasthead.propTypes = {
```
Any way to get these props to show in the storybook?

jeff-phillips-18 commented on src/components/VerticalNavigation/__mocks__/mockIconBarChildren.js:
```diff
@@ -0,0 +1,44 @@
+        <span className="badge badge-pf-bordered" />
+      </a>
+    </li>
+    <li className="dropdown">
```
This should use the Dropdown component, the menu items do not show currently

mturley replied:
Yeah, this was just thrown in for the demo yesterday. I'll use the right components.

jgiardino replied:
It looks like this masthead is included here as just a placeholder. If so, then I'm wondering how much effort we should put into creating the mock masthead, especially if the plan is to actually build a masthead component later.

jeff-phillips-18 commented on src/components/VerticalNavigation/__mocks__/mockIconBarChildren.js:
```diff
@@ -0,0 +1,44 @@
+    </li>
+    <li className="dropdown">
+      <a
+        href="#0"
```
This `href=#0` wreaks havoc and sends me off to another component, messes up the storybook navigation.

jgiardino replied:
I don't see that behavior, but I didn't build this locally.

jeff-phillips-18 commented on src/components/VerticalNavigation/VerticalNavigationItem.js:
```diff
@@ -0,0 +1,443 @@
+    }
+    this.setActive();
+    onClick && onClick(primary, secondary, tertiary);
+    // TODO other item-level nav props? href? route?
```
Selecting a leaf node should dismiss the menus.

mturley replied:
Ah, good to know.

mturley commented:
@priley86, @jeff-phillips-18, thanks for all the feedback.

jgiardino commented on sass/patternfly-react/_patternfly-react.scss:
```diff
@@ -2,3 +2,7 @@
+
+.navbar-brand-txt {
+  line-height: 34px;
+}
```
I'm not able to find where this is used in the component. If we are setting this value, can we use a patternfly variable? In [navbar-vertical.less](https://github.com/patternfly/patternfly/blob/master/src/less/navbar-vertical.less#L102), a similar value of 35px is set using `@navbar-pf-vertical-navbar-brand-min-height;`

mturley replied:
Without this style line, the "Patternfly React" title text was just rendering off-center vertically. I added this to visually match the Patternfly Ng title style. I'll use the variable, I didn't see that.

mturley replied:
Added a PR to core for this: https://github.com/patternfly/patternfly/pull/922

jgiardino commented on src/components/VerticalNavigation/VerticalNavigationBrand.js:
```diff
@@ -0,0 +1,41 @@
+      {titleImg && (
+        <img className="navbar-brand-name" src={titleImg} alt={title} />
+      )}
+      {!titleImg && title && <span className="navbar-brand-txt">{title}</span>}
```
Oh, this is where that class is being used. Is this variation viewable in the storybook?

mturley replied:
@jgiardino, yes, this is the "Patternfly React" text that appears at the top in all the Vertical Nav stories except the last one (custom masthead). If you render a `<VerticalNavigation.Masthead title="..." />` it will use that span.navbar-brand-txt

mturley replied:
Ah, I just realized that the link I had posted hid the storybook navigation so you couldn't switch to the other stories.  Sorry about that!  You can see them all at https://mturley.github.io/patternfly-react/

jgiardino commented:
Hey Mike, I'm glad to see you were able to get a storybook link posted :-) There's only one link, and I don't see a way to view other variations. Were there other variations of the navigation available in the storybook?

There are only a couple of issues that I noticed regarding the interaction:

- The list items in the navigation are rendered with just `<a>` and no `href` attribute, which prevents keyboard focus on the menu item.
- At smaller screen sizes (e.g. when `.show-mobile-nav` gets applied), the menu is no longer persistent. For example, in the [pf core test page](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/vertical-navigation-primary-only.html#0), when I click the hamburger, then click a menu item, the menu should be hidden again. I don't see this behavior in the storybook example.

@Rohoover Could you also take a look and see if there's anything I missed?

jeff-phillips-18 commented:
@jgiardino If you use https://mturley.github.io/patternfly-react you can see the different examples and you'll see the navigation issue I mentioned above.

jgiardino commented:
Thanks, @jeff-phillips-18.

Also, I just updated the 2nd bullet in my previous comment about hiding the secondary and tertiary menus after clicking a final destination.

Rohoover commented:
The one thing I'm noticing is that when you click on an item, the nav remains, and in the site I usually use for comparison: 
https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/vertical-navigation-with-tertiary-pins.html#0
The nav closes after click.

serenamarie125 commented:
@Rohoover I noticed that as well, but it is consistent with the Angular PF behavior.  @jeff-phillips-18 do we know which is correct?

jeff-phillips-18 commented:
@serenamarie125 I believe it should dismiss on click. The core patternfly example should be considered the standard.

serenamarie125 commented:
So @jeff-phillips-18 I think the reason it remains viewable, is so you can select the PIN feature

jgiardino commented:
Like @jeff-phillips-18, I also thought the patternfly core example was the standard we should follow. I also didn't realize that the angular implementation was different. 

@mcarrano Can you weigh in on this one?

I did see that @mturley made updates based on our previous comments. But now these updates are slightly different in interaction behavior with the core patternfly example:

Steps to reproduce:
1. Hover over a menu item that has a submenu to show the submenu. 
2. Click that same menu item 

**Results in the pf core example:** 
- The submenu remains visible
- The clicked menu item displays as Active*. 

**Results in the latest [pf-react implementation](https://mturley.github.io/patternfly-react/?selectedKind=Vertical%20Navigation&selectedStory=Flat%20%28Inline%20JSX%29&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs):** 
- The submenu is hidden
- The clicked menu item displays as Active*. 

***Note:** When the submenu of an Active menu item is visible, an arrow icon displays at the top of the submenu. This icon lets the user pin the menu

mcarrano commented:
Regarding your question @jgiardino , yes, if the user clicks on a menu item that is a parent (i.e. has further submenus) the submenu should remain open. This is because typically there would not be an intermediate landing page and therefore no place to go.

In looking into this more deeply, I also see how the behavior of Angular and Core differ when the user clicks the last node in the tree to navigate to a new page. The advantage of the Angular implementation seems to be that keeping the menu open facilitates pinning. Was this the motivation for doing it this way @serenamarie125 and @jeff-phillips-18 ? If that's the case the Core example (and React) should both be made consistent. I think this is only an issue if pinning is used. Thoughts?

priley86 commented on src/components/VerticalNavigation/VerticalNavigation.js:
```diff
@@ -0,0 +1,482 @@
+
+    return (
+      <NavContextProvider
+        idPath={'/'}
```
should this be a defaultProp?

mturley replied:
This could be a defaultProp on NavContextProvider, but I was passing it explicitly because the only place where the path should be `'/'` is at the root, in the idPath passed to primary items. If it were a default prop and i forgot to wire up the paths correctly in a sub-item's context provider, i might get `'/'` starting somewhere down the tree and wonder why my paths are truncated, rather than just getting an error.

priley86 replied:
a defaultProp on `VerticalNavigation.js` rather @mturley ? `VerticalNavigation.defaultProps = { idPath: '/' }`

mturley replied:
But you wouldn't ever pass it as a prop into VerticalNavigation. That idPath should only ever be '/' in the top-level provider.  I could have instead made it a defaultProp on VerticalNavigationItem, but i'm not sure how context and defaultProps play together.

priley86 replied:
Ok, I am just trying to map existing understanding from React-Router...this works for now ;)

mturley replied:
No worries! I might move this value somewhere else, or align it more with react-router later on when we add that support.

priley86 commented on src/components/VerticalNavigation/VerticalNavigation.js:
```diff
@@ -0,0 +1,482 @@
+  }
+
+  navigateToItem(item) {
+    console.log('TODO NAVIGATE!', item); // TODO
```
should this just call a callback function if one exists for now? fine with leaving it open ended for v1 if others are...

mturley replied:
Yeah, i forgot I hadn't done that already. I'll fill it in with an onNavigate callback you can use, but the item components themselves also have onClick and href props that can be used instead.

mturley commented:
@jgiardino and @jeff-phillips-18 thanks for the bug reports, and @priley86 thanks for the code feedback. I'll take all this into my final round of changes.

mturley commented:
@mcarrano to your comment, it sounds like we should use the behavior of leaving the nav open when `pinnableMenus` is true, and closing them otherwise. Does that sound right to everyone? Should we expose another prop `closeOnSelection` or something instead?

priley86 commented on src/components/VerticalNavigation/VerticalNavigationBrand.js:
```diff
@@ -0,0 +1,47 @@
+    </span>
+  );
+  return href ? (
+    <a href={href} className="navbar-brand">
```
would be nice in the future to maybe expose the `onClick` of the `a` too... @jeff-phillips-18 do you have any thoughts? I'm fine either way for now...

mcarrano replied:
Yes, leaving nav open when menus are pinnable makes sense to me @mturley . If everyone agrees with this approach, I can file an issue to make the Core behavior consistent.

jeff-phillips-18 replied:
For the brand icon/text, I agree it should be able to use an href, an onClick callback, or nothing at all.

mturley replied:
Ok I'll make sure it is flexible.

priley86 commented on src/components/VerticalNavigation/VerticalNavigationTertiaryItem.js:
```diff
@@ -0,0 +1,28 @@
+  BaseVerticalNavigationTertiaryItem
+);
+
+VerticalNavigationTertiaryItem.propTypes =
```
I seem to see the `propTypes` extensions being sometimes referenced directly vs using the `...` destructuring assignment. Can we just be consistent and use the destructuring assignment everywhere... ie:

```
VerticalNavigationTertiaryItem.propTypes = { ...BaseVerticalNavigationTertiaryItem.propTypes;
}
```

mturley replied:
Yeah, agreed. I meant to use the destructuring everywhere, if I missed it anywhere else I'll fix it.

mturley commented:
Okay @priley86, @jeff-phillips-18, @serenamarie125, @patternfly/patternfly-react-maintainers. This one is finally ready for a final review and merge. Tying up loose ends turned into an all-nighter... but now it's done. DONE I tell you.

mcarrano commented:
@mturley @jgiardino When I test the version on Storybook, I see that the nav still closes when I click an item that has children. Hadn't we decided that the nav should remain open in those cases and only clicking on the bottom node in the hierarchy would close the menu?

jgiardino commented:
@mcarrano The latest implementation only persists the menu if the option to pin the menu is available. Otherwise the menus are dismissed on click. Is this difference in behavior consistent with what you meant by [this comment above](https://github.com/patternfly/patternfly-react/pull/88#issuecomment-359497808), or were you thinking something else?:
> I think this is only an issue if pinning is used.

jgiardino commented:
I did notice a minor issue :-( 
Steps to reproduce are:
1. Go to [the pinnable example](https://mturley.github.io/patternfly-react/?selectedKind=Vertical%20Navigation&selectedStory=Pinnable%20Menus&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)
2. Click the hamburger to show icons only
3. Click a primary menu item
4. Click the icon to pin the menu
**Results:** The contents area does not resize to accommodate the width of the pinned menu.
Complete these steps on the [patternfly test page](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/vertical-navigation-with-tertiary-pins.html#0) for an example of what should happen.

mcarrano commented:
@jgiardino Sorry if my earlier comment was unclear, but there were really two separate issues. Yes, there was the question of keeping the menu open on click if pinning is enabled.  But there was also a second question as to what should happen when the user clicks on a menu item that has a sub-menu. Steps to reproduce:

    Hover over a menu item that has a submenu to show the submenu.
    Click that same menu item

Right now, this dismisses the menu which assumes that the parent item has a landing page. In the original PatternFly core implementation, the menu is only dismissed when the user clicks on the item in the last submenu to trigger navigation to that target page.

Does that clarify?

mturley commented:
@mcarrano that makes more sense, I think @jgiardino and I both thought you were referring to the same pinning issue. I'll take a look at that.

Also @jgiardino that is certainly a problem, good catch!

Thanks guys. I'll push follow-up changes in a little bit.

amirfefer commented:
Great progress! looks great :)
some feedback:
In mobile state, there is no `back` button, as you can see in the pf demo page:
![screenshot from 2018-01-24 19-55-10](https://user-images.githubusercontent.com/11807069/35349564-f63b86dc-0143-11e8-8692-e1d16f0f4130.png)
while it's missing here: 
![screenshot from 2018-01-24 20-14-25](https://user-images.githubusercontent.com/11807069/35349607-174bef74-0144-11e8-8a17-d8a5e09b0ed2.png)

When refreshing the page, (or by routing to a different page), the vertical navigation returns to be expanded mode even if before it was collapsed. In the pf demo page, that state is preserved through session storage.  

Could we make the menu to respond immediately after a click? we have a [feedback](http://projects.theforeman.org/issues/21566) on pf vertical nav which says that responding delay sometimes is too long.

mturley commented:
@amirfefer thanks for your feedback.

1. That's interesting, I think it must only be showing those when `pinnableMenus` is true, since those buttons are used for pinning too on desktop. Until recently `pinnableMenus` defaulted to true, so all the examples had working back buttons on mobile. If you try the Pinnable Menus example in mobile form it has the back buttons. So that's a bug I need to fix.

2. Hmm. Persisting state info through a page reload isn't something I've seen elsewhere in patternfly-react yet. As it is now, the app could drive this state and control the vertical nav through its special controlled state props. But I wonder if we want to add a fallback default behavior that does this automatically, like the other overridable internal state. Since this one can be lifted to the application though, I wonder if that's good enough.

3. When you say "respond immediately after a click", do you mean clicking on a sub-menu to force its hover delay to end? Clicks on mobile are already immediate, but yeah I agree it would be nice to be able to click and skip that delay if you want to move around quickly. However, is this a behavior that should be fixed in the pf core and all implementations? Or would it be okay to include this extra functionality just in patternfly-react and make an issue for putting it in other repos? @priley86 do you have thoughts on this?

mturley commented:
@jgiardino I noticed another issue similar to yours-- when you close the menu with the hamburger button, the body content doesn't move left and you just get a big white margin. I need to clean up the handling of dynamic body classes.

jgiardino commented:
@mcarrano Thanks for clarifying and checking all the examples! This issue is only present in the non-pinnable examples. I agree with you. Just to make sure we're on the same page, these are the steps we're referring to and the expected behavior:

Use case: vertical nav has more than one level, and the submenus cannot be pinned
1. Hover over a primary menu item that has a submenu
2. When the submenu is visible, click the primary menu item

**Expected result:** 
- the submenu continues to display
- the primary menu item displays as the active menu item in the primary menu
    - Is this correct? This is the behavior in core, the angular components don't show a non-pinnable example

**Current result:** the submenu is hidden

mcarrano commented:
Yes. What you describe above is what I expect @jgiardino

jeff-phillips-18 commented:
In angular-patternfly, selecting a parent item selects its first child. If that child has children, it selects its first child, etc.

mcarrano commented:
Are you saying that it directly navigates to that page @jeff-phillips-18 or just pre-selects them in the menu and keeps the nav open?

priley86 commented:
@mturley @jeff-phillips-18  re: # 3 hover delay/mouse click, IIUC this may just be an issue with the older vertical nav implementation in Core being explicitly set to `500ms` hover delay? It seems like we just have some differing UX requests (looking at @amirfefer and @Rohoover comments in the issue posted).

At any rate, I think it is nice providing `hoverDelay`/`hideDelay` as props and defaulting the delay to `500ms` like you have done. Your also updating the nav on clicks and hovers ([here](https://github.com/mturley/patternfly-react/blob/b7fbab22ab5e395f7570348cbd47074d45d7a206/src/components/VerticalNav/VerticalNavItemHelper.js#L197) and [here](https://github.com/mturley/patternfly-react/blob/b7fbab22ab5e395f7570348cbd47074d45d7a206/src/components/VerticalNav/VerticalNavItemHelper.js#L248)) , so maybe a nice addition to support both use cases if the consumer really wants to just force clicks and ignore hovers is to *not* update the nav if the `hoverDelay` prop is set zero? This just seems like a decent compromise, but I definitely support a consistent UX pattern in products...

jeff-phillips-18 commented:
@mcarrano Yes, it navigates to that page (or calls the active callback for the final leaf node). 

@mturley The addition of the 'active' class to any item should correspond to the current navigation state. So selecting an intermediate item should only select its parent if also selecting some final leaf node (which the application then navigate to).

mturley commented:
@jeff-phillips-18 Okay, in that case we do need to select the leaf nodes on a parent click. Because I don't think there is a way to express the style of an intermediate item being selected without having its parent also be active (at least I seemed to run into that while testing).

So are you saying we should not be able to select only a secondary item which has children (and not the primary or tertiary items), and clicking that will always select the leaf item? Or we need to add a mechanism for selecting non-leaf items? without setting the "active" class on the items in its subtree? Sounds like the former.

mcarrano commented:
In looking at the core implementation, it looks like clicking on a parent node also pre-selects the first child, but the menu does not close. From looking at the example it's impossible to tell whether this also triggers navigation to the first child, but I can try to get some clarification on that. My concern with the approach that @jeff-phillips-18 describes for Angular implementation is whether this will be a concern for mobile where I need to click on an item to open child menus. Thoughts?

jgiardino commented:
+1 for getting clarification on the current or expected behavior, @mcarrano. If we could reach out to anyone working on a project that currently uses the angular vertical nav, that would be great, too. (Although I am assuming that behavior is built into the nav component, and would be consistent across products.) None of the patternfly examples actually load new pages based on nav selection, so it's hard to know for sure what the expected behavior would be from looking at our examples.

That's also a good point about the mobile experience. We really need a giant matrix to capture all the variations and expected behaviors for different actions.

jeff-phillips-18 commented:
There is no 'Active' state set on anything unless there is navigation. Items are 'highlighted' only due to hover or to being 'active'.

Mobile works differently. Clicking on a parent displays its submenu, no navigation occurs. Navigation only occurs when clicking on a leaf node.

mcarrano commented:
I raised this issue during the Pf-UXS meeting today and the consensus was that the React implementation should follow PF Core in regarding to behavior when clicking on parent nodes, i.e. first child is selected, but menu remains open (and navigation does not happen until the leaf node in clicked). 

@serenamarie125 please feel free to comment further.

mturley commented:
@mcarrano, to @jeff-phillips-18's point that is a bit of a problem. Active state corresponds to where you are navigated, so we can't select the first child without navigating to it.

Is that what PF core does, is add the "active" class to the first child, while there might also be an "active" class on the current page's nav item?  I wrote the React implementation with a single activePath, so there can't be two different paths down the item tree with active classes in them.

mturley commented:
To clarify, also as @jeff-phillips-18 said, mobile is different, we also have a `mobilePath` that is separate from the `activePath`.  We would have to add something like `clickedPath`... or maybe just make it so when clicking the parent, nothing happens other than skipping its hover timer to show the hover effect immediately?  That accomplishes the same goal as selecting the first child, yes?

I need to take a closer look at pf core.

jeff-phillips-18 commented:
Correct @mturley, the core implementation never actually navigates, it just wasn't necessary to add actual navigation into the app. The intent (and I implemented both the core and angular-patternfly versions) was that it would navigate. Active indicates current navigation path. 

There is no selecting items without navigation. So selecting the first child on a parent click would mean navigating to that item (or its first child(.

mturley commented:
Gotcha @jeff-phillips-18. So it sounds like what the click on a parent item should do is, just skip the hover timer. Unless there is an href or an onClick directly on that item.

mcarrano commented:
@mturley The interaction issue I raised above was further discussed during today's Reop Leads meeting with @jeff-phillips-18 and @jgiardino present. Consensus is to implement to match Anglular behavior as @jeff-phillips-18 describes. We may need to revisit this from an accessibility and usability perspective in the future, but for now we will prioritize consistency.

mturley commented:
Sounds good. I am definitely in favor of usability/a11y improvements after this PR, but yes this PR's goal is definitely just to get react caught up with where angular is today.

mturley commented:
Thanks for helping me figure this out @mcarrano @jeff-phillips-18

mturley commented:
@jeff-phillips-18 @priley86, okay guys this is ready for another, hopefully final, review.

I squashed the whole PR into two commits: changes before the latest discussion and changes since. So if you want to see what's new, look at https://github.com/patternfly/patternfly-react/pull/88/commits/711053c351d62cefe8f4631cdcaf54cda0d8182f.

Changes in this commit:
* Refactor hoverTimer so it can be skipped, and use that to skip it on item click
* Lift all the remaining item state up to the VerticalNav container (`hoverPath`, `pinnedPath`)
* Fix bugs around body classes (was causing the bad element positioning)
* Add a body click handler that will skip the blur timer\*
* Add `hoverDisabled` and `blurDisabled` props to allow further customization
* Misc refactoring to reduce code duplication

\* Note, the one weird thing here is that now, when the main body is clicked, the hover state will be cleared. This is consistent with many nav implementations out there, and I think is necessary to complement the click-to-skip-hover-timer behavior. Basically, if you want to be able to click to get around the timer, this means you can do so everywhere. And if you use the `blurDisabled` prop, it is the only way aside from clicking another item to hide the hovered nav.

The click-body-to-blur behavior isn't in core, but neither is click-item-to-skip-hover-timer. I think if we're adding one, we should add both.

mturley commented:
Just updated my storybook at https://mturley.github.io/patternfly-react/.

priley86 commented on src/components/VerticalNav/VerticalNav.js:
```diff
@@ -4,14 +4,20 @@ import cx from 'classnames';
 import VerticalNavMasthead from './VerticalNavMasthead';
-import { bindMethods, filterChildren, findChild } from '../../common/helpers';
+import {
+  bindMethods,
```
open question here - would anyone be opposed to exposing/consuming all of our helpers like `bindMethods` and the others you've added in this PR from `src/index.js`? It would be nice to use these downstream as well I am thinking, but will see how others feel...

priley86 commented on src/common/helpers.js:
```diff
@@ -29,6 +29,9 @@ export const findChild = (children, validator) => {
   return array && array.find(validator);
 };
 
+export const propsChanged = (propNames, oldProps, newProps) =>
```
nice!

mturley commented:
@jeff-phillips-18 @priley86 @amirfefer @mcarrano @jgiardino:

This should now be all ready to go. I just pushed a few more improvements, because I realized I had missed one thing above (persisting collapsed/pinned state to sessionStorage). I squashed them into the second commit again, so to see the changes since the last discussion, that commit is now https://github.com/patternfly/patternfly-react/pull/88/commits/0272ce9d5cc41d5d217bd33208864bd254de0e84

As a bonus, I implemented the sessionStorage enhancement as part of the `controlled()` higher-order component. So it is not nav-specific: we can now easily persist state from any component to sessionStorage using controlled()'s new persist option.

Reading and re-reading this whole thread, I am pretty confident that I now have addressed everything we discussed. Please let me know if there's anything else!  Looking forward to putting this one behind me.

Thanks again guys. It's happening.

mturley commented:
Pushed one more commit to export the helpers per @priley86's suggestion :)

mturley commented:
Okay. Straggler commits were needed to fix some errors. Squashed everything into two commits again. Changes since last discussion are in https://github.com/patternfly/patternfly-react/pull/88/commits/459c6b33ed1e2c82efa1c31f83131ee71ef44957 now.

mturley commented:
Properly fixed the hover bug, as @jeff-phillips-18 pointed out my fix from before didn't cover all cases. I did a more thorough test and I'm not finding any more issues.

priley86 commented:
great job @mturley. I think this was a very challenging component and you broke it down into pieces nicely (also, i really appreciate the helpers added). I am comfortable merging this PR at this stage if others are and like other components, we can continue to iterate as necessary. It will be nice to see how it goes on the integration end and hopefully folks will report back their feedback.

👍 major kudos for tackling this one and building the robust PF wizard!!!

mturley commented:
Thanks @priley86. I pushed a few more minor tweaks from discussions with @jeff-phillips-18 (fixed corner cases around hover/blur timers). I guess that invalidated your approval. If you could re-approve when you get a chance that'd be great!

priley86 commented:
@mturley nice job on the fixes. Again I am OK to merge here. Another thing I noted with the `controlled` pattern here - we may be able to connect that to a redux-session or a redux local storage as an enhancement in the future. This works great for now though!

https://egghead.io/lessons/javascript-redux-persisting-the-state-to-the-local-storage

mturley commented:
Thanks for the reviews guys. @priley86 I like the idea of using redux patterns for local storage, I do think though that even if we add that support, it should still be able to fall back on window.sessionStorage to support users who aren't using redux.

priley86 commented:
@mturley sure - i think redux is an optional replacement for the consumer. valid point 👍

serenamarie125 commented:
@jeff-phillips-18 @mturley just want to confirm that the strategy for this component was to match the core implementation, which can be found here: https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/vertical-navigation-with-tertiary-pins.html

There is a difference between the Pinning implementation between core & angular.  This PR matches the core implementation.  Just noting here :)

jeff-phillips-18 commented:
Found a few issues with mobile mode.

1) if I collapse (only icons) then go into mobile mode, the primary menu should show both icons and text
2) if I collapse (only icons) then go into mobile mode, the tertiary menu gets displayed incorrectly
![image](https://user-images.githubusercontent.com/11633780/35683054-bf1ecff4-0730-11e8-9bb4-8fb3a6d03c3d.png)
3) In mobile mode, the body content should have no left margin.

mturley commented:
Fixed the mobile issues @jeff-phillips-18 was seeing.

mturley commented:
Aww every time I rebase on upstream master it dismisses all my shiny green approvals... @priley86 @serenamarie125 :)

jeff-phillips-18 commented:
🎉

mturley commented:
![avoskic](https://user-images.githubusercontent.com/811963/35756011-95e547a0-0837-11e8-86fb-1405a78cdeca.gif)
